homeassistant/components/monzo/__init__.py

```diff
@@ -0,0 +1,220 @@
+    # Get sensors as defined in configuration.yaml
+    sensors = hass.data.get(DATA_MONZO_CONFIG, {}).get(CONF_SENSORS, {}).get(
+        CONF_MONITORED_CONDITIONS, list(SENSORS))
+
+    monzo = MonzoObject(hass, sensors, config_entry)
+    await monzo.async_update()
+
+    # Make Monzo client available
+    hass.data[DOMAIN][DATA_MONZO_CLIENT][config_entry.entry_id] = monzo
+
+    _LOGGER.debug("proceeding with setup")
+
+    # Create config entries for each component.
+    for component in ('sensor',):
+        hass.async_create_task(hass.config_entries.async_forward_entry_setup(
+            config_entry, component))
+
+    async def refresh_sensors(event_time):
+        """Refresh Monzo data."""
+        _LOGGER.debug('Refreshing Monzo data')
+        await monzo.async_update()
+        async_dispatcher_send(hass, TOPIC_UPDATE)
+
+    # Automatically poll Monzo on set interval and update stored data
+    hass.data[DOMAIN][DATA_MONZO_LISTENER][
+        config_entry.entry_id] = async_track_time_interval(
+            hass, refresh_sensors,
+            hass.data.get(DATA_MONZO_CONFIG, {}).get(CONF_SCAN_INTERVAL,
+                                                     DEFAULT_SCAN_INTERVAL))
+
+    _LOGGER.debug("async_setup_monzo is done")
+
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload a Monzo config entry."""
+    for component in ('sensor',):
+        await hass.config_entries.async_forward_entry_unload(
+            config_entry, component)
+
+    hass.data[DOMAIN][DATA_MONZO_CLIENT].pop(config_entry.entry_id)
+
+    remove_listener = hass.data[DOMAIN][DATA_MONZO_LISTENER].pop(
+        config_entry.entry_id)
+    remove_listener()
+
+    return True
+
+
+class MonzoObject:
+    """Define a generic Monzo object."""
+
+    def __init__(self, hass, sensor_conditions, config_entry):
+        """Initialize."""
+        from monzo.monzo import Monzo
+        from monzo.auth import MonzoOAuth2Client
+
+        self.data = {}
+        self._hass = hass
+        self.sensor_conditions = sensor_conditions
+        self.config_entry = config_entry
+
+        client_id = config_entry.data['tokens'][CONF_CLIENT_ID]
+        client_secret = config_entry.data['tokens'][CONF_CLIENT_SECRET]
+        access_token = config_entry.data['tokens'][CONF_ACCESS_TOKEN]
+        refresh_token = config_entry.data['tokens'][CONF_REFRESH_TOKEN]
+        last_saved_at = config_entry.data['tokens'].get(CONF_LAST_SAVED_AT)
+        expires_at = config_entry.data['tokens'].get(CONF_EXPIRES_AT,
+                                                     last_saved_at)
+
+        oauth_client = MonzoOAuth2Client(
+            client_id=client_id,
+            client_secret=client_secret,
+            access_token=access_token,
+            refresh_token=refresh_token,
+            expires_at=expires_at,
+            refresh_callback=self.update_config_entry)
+
+        self.client = Monzo.from_oauth_session(oauth_client)
```

Comment:
Is this doing any I/O ? It's instantiated within an async context so it's not allowed to do any I/O